feat(cms): add copyPostContent field and view for post content duplication

ericsen-tsai opened the pull request:
## Summary

This PR adds a new feature to the CMS admin UI that enables users to copy post content in a formatted markdown-like structure. The feature appears as a sidebar field in the post item view, allowing content editors to easily duplicate and share post content with proper formatting.

## Changes

- Added `copyPostContent` virtual field to the Post list configuration in `packages/cms/lists/post.ts`
  - Field is displayed in the sidebar of the item view
  - Hidden in create and list views
- Created new React component `packages/cms/lists/views/copy-post-content.tsx` (240 lines)
  - Provides checkbox UI for selecting which post fields to include in the copy
  - Supports 8 field types: title, subtitle, brief, content, ogDescription, opening, postChoiceQuestions, postEssayQuestions
  - Fetches post data via GraphQL query
  - Formats Draft.js content (brief, content) to plain text using `convertFromRaw`
  - Formats choice questions with markdown checkboxes (checked for correct answers)
  - Formats essay questions with hints
  - Uses Clipboard API to copy formatted content
  - Shows loading and error states
  - Provides visual feedback when content is copied ("已複製" for 2 seconds)
  - All fields are selected by default
- Updated GraphQL schema in `packages/cms/schema.graphql` to include the new `copyPostContent` field

## Additional Notes

- The component handles errors gracefully with try-catch blocks for Draft.js conversion and clipboard operations
- The copy button is disabled when no fields are selected, during loading, on error, or when post data is unavailable
- The formatted output uses markdown-style headers (##) for section separation
- Choice questions display options as markdown checkboxes with `- [x]` for correct answers and `- [ ]` for incorrect ones

## Screenshot(s)

https://github.com/user-attachments/assets/67e26e5d-97ad-4b2a-96ec-10ae5f6cdfd1





## Reference(s)

- [Notion Page](https://www.notion.so/twreporter/2cd8dbdca93280ed81cec48caf67ce6b?v=1588dbdca93281dfa43c000c8415265e&source=copy_link)

ericsen-tsai committed:
feat(cms): add copyPostContent field and view for post content duplication

- Introduced a new virtual field `copyPostContent` in the Post list configuration to facilitate content duplication.
- Added a corresponding React component for the UI, allowing users to select and copy specific fields from a post.
- Implemented a GraphQL query to fetch post details for the duplication feature.

nickhsine commented on packages/cms/lists/views/copy-post-content.tsx:
```diff
@@ -0,0 +1,240 @@
+    })
+  }
+
+  const formatContent = useCallback(() => {
```
question:
想問問這邊會需要 `useCallback` 嗎？

ericsen-tsai replied:
The `useCallback` is redundant because it's only used for `handleCopy` function
I removed it in [fix(cms): enhance post content copying functionality](https://github.com/kids-reporter/kids-reporter-monorepo/pull/872/changes/d8620262f501cb6c8e1c63aacab35c9218233a2b).

nickhsine commented on packages/cms/lists/post.ts:
```diff
@@ -521,6 +521,21 @@ const listConfigurations: ListConfig<any> = list({
+        graphql.field({
+          type: graphql.JSON,
+          async resolve(item: Record<string, any>) {
+            return { postId: item.id }
```
question:
不曉得是否直接 `return item` ？
如此一來，view 的部分就不需要再次打 GQL 拿 post 資料？

ericsen-tsai replied:
Sure, adjusted in [refactor(cms): fetch post content data in a single query](https://github.com/kids-reporter/kids-reporter-monorepo/pull/872/changes/eacd06e57712b8edb035220bcac0bd89651bbae8)

nickhsine commented:
LGTM 👍 
兩個疑問和 copilot 的留言再麻煩看看。

ericsen-tsai committed:
fix(cms): enhance post content copying functionality

- Updated the `copy-post-content` component to improve type safety by defining `PostChoiceQuestion` and `PostEssayQuestion` types.
- Refactored the `formatContent` function to remove unnecessary use of `useCallback`.
- Implemented cleanup logic for the timeout used in the copy feedback mechanism.
- Enhanced checkbox rendering for field selection with unique IDs for accessibility.

ericsen-tsai committed:
refactor(cms): fetch post content data in a single query

ericsen-tsai commented on packages/cms/lists/views/copy-post-content.tsx:
```diff
@@ -0,0 +1,240 @@
+    }
+
+    return parts.join('\n\n')
+  }, [post, selectedFields])
```
Fixed in [fix(cms): enhance post content copying functionality](https://github.com/kids-reporter/kids-reporter-monorepo/pull/872/changes/d8620262f501cb6c8e1c63aacab35c9218233a2b)

ericsen-tsai commented on packages/cms/lists/views/copy-post-content.tsx:
```diff
@@ -0,0 +1,240 @@
+    ) {
+      parts.push('## 思辨題組')
+      post.postEssayQuestions.forEach((q: any, index: number) => {
+        parts.push(`\n### ${index + 1}. ${q.title}`)
```
Fixed in [fix(cms): enhance post content copying functionality](https://github.com/kids-reporter/kids-reporter-monorepo/pull/872/changes/d8620262f501cb6c8e1c63aacab35c9218233a2b)

ericsen-tsai commented on packages/cms/lists/views/copy-post-content.tsx:
```diff
@@ -0,0 +1,240 @@
+            />
+            <CheckboxLabel>{fieldLabels[field]}</CheckboxLabel>
+          </CheckboxContainer>
+        ))}
```
Fixed in [fix(cms): enhance post content copying functionality](https://github.com/kids-reporter/kids-reporter-monorepo/pull/872/changes/d8620262f501cb6c8e1c63aacab35c9218233a2b)

ericsen-tsai commented on packages/cms/lists/views/copy-post-content.tsx:
```diff
@@ -0,0 +1,240 @@
+    try {
+      await navigator.clipboard.writeText(content)
+      setCopied(true)
+      setTimeout(() => setCopied(false), 2000)
```
Fixed in [fix(cms): enhance post content copying functionality](https://github.com/kids-reporter/kids-reporter-monorepo/pull/872/changes/d8620262f501cb6c8e1c63aacab35c9218233a2b)

nickhsine commented on packages/cms/lists/post.ts:
```diff
@@ -521,6 +521,46 @@ const listConfigurations: ListConfig<any> = list({
+        createView: { fieldMode: 'hidden' },
+        itemView: { fieldPosition: 'sidebar' },
+        listView: { fieldMode: 'hidden' },
+      },
```
suggestion:
`copyPostContent` 單純是編輯介面需要使用，
可以加上 `graphql: { omit: true }` 來 disable GQL query。

ericsen-tsai replied:
Fixed in [fix(cms): omit copy post content field in graphql schema](https://github.com/kids-reporter/kids-reporter-monorepo/pull/872/changes/eba84a835fcf4d197e06c2f0c0fb79f2d5718928)

ericsen-tsai replied:
I found that omitting all will cause this virtual field disappear in UI.
I merged another PR to address this.
https://github.com/kids-reporter/kids-reporter-monorepo/pull/874

ericsen-tsai committed:
fix(cms): omit copy post content field in graphql schema
